packages/remirror__extension-react-format-code/src/format-code.tsx

```diff
@@ -0,0 +1,110 @@
+import React, { CSSProperties } from 'react';
+import type { FindProsemirrorNodeResult } from '@remirror/core';
+import { findParentNodeOfType, isElementDomNode } from '@remirror/core';
+import {
+  defaultAbsolutePosition,
+  hasStateChanged,
+  isPositionVisible,
+  Positioner,
+} from '@remirror/extension-positioner';
+import { PositionerPortal } from '@remirror/react-components';
+import { useCommands, useRemirrorContext } from '@remirror/react-core';
+import { usePositioner } from '@remirror/react-hooks';
+
+export interface CodeBlockFormatProps {
+  className?: string;
+  text?: string;
+  offset?: { x: number; y: number };
+}
+
+interface CodeBlockPositionerData {
+  codeBlockResult: FindProsemirrorNodeResult;
+}
+
+export const CodeBlockFormatCode = ({
+  className = '',
+  text = 'Format',
+  offset = { x: 0, y: 0 },
+}: CodeBlockFormatProps): JSX.Element => {
+  const position = usePositioner(createPositioner, []);
+  const { getState } = useRemirrorContext();
+  const { focus, formatCodeBlock } = useCommands();
+  const { selection } = getState();
+
+  const nodeLanguage = findParentNodeOfType({
+    selection,
+    types: 'codeBlock',
+  })?.node.attrs.language;
+
+  return (
+    <PositionerPortal>
+      <button
+        ref={position.ref}
+        className={className}
+        onClick={() => {
+          formatCodeBlock();
+          focus();
+        }}
+        style={
+          {
+            '--x': `${position.x + offset.x}px`,
+            '--y': `${position.y + offset.y}px`,
+            display: !nodeLanguage ? 'none' : 'block',
+          } as CSSProperties
+        }
+      >
+        {text || 'Format'}
+      </button>
+    </PositionerPortal>
+  );
+};
+
+function createPositioner(): Positioner<CodeBlockPositionerData> {
```

Comment:
This is the [same positioner as in \@remirror/extension-react-language-select](https://github.com/remirror/remirror/blob/1c3da950e55d21f2406fe844819c628b10ab5eac/packages/remirror__extension-react-language-select/src/language-select.tsx#L202). I'd like for the two components to share the code, but I'm not sure where to move the function. Advice?